### PR TITLE
kubeadm: Secure apiserver -> kubelet communication and set storage backend to etcd3

### LINKED ISSUE #118: Make the apiserver -> kubelet communication secure
Generate these files in the cert phase as well:
 - kubelet-ca.crt
 - kubelet-ca.key
 - apiserver-kubelet.crt
 - apiserver-kubelet.key

Pass to apiserver:
 - `--kubelet-certificate-authority=kubelet-ca.crt`
 - `--kubelet-client-certificate=apiserver-kubelet.crt`
 - `--kubelet-client-key=apiserver-kubelet.key`

Pass to kubelet:
 - `--client-ca-file=kubelet-ca.crt`

**Difficulty:** How do we transfer the `kubelet-ca.crt` file from master to node?
Can we somehow transfer two ca:s with a kubeconfig and the token JWS flow...?

Or do we have to extend our custom Go struct for now to pass two CA certs until we have something like kubeconfig v2?

Thoughts? @jbeda @pires

@mikedanese Did I get it right?

### LINKED ISSUE #129: Set the default storage backend to etcd3
We already deploy etcd 3.0.14 for kubeadm, we should follow that up with actually using etcd3 mode in the apiserver, otherwise we don't gain much.

This is pretty easy to do, we should just set `--storage-backend=etcd3` on the apiserver

### PR DESCRIPTION
**Pull request by luxas**
**What this PR does / why we need it**:

Switch storage backend to etcd3
Writes ca.crt down to disk so we can set `--client-ca-file=/etc/kubernetes/ca.crt` for the kubelet.
Adds --kubelet-client-{certificate,key} to the apiserver args and make it allowed to access the kubelets
In some cases the `path` dependency is switched to `filepath`

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*:

fixes https://github.com/kubernetes/kubeadm/issues/118
fixes https://github.com/kubernetes/kubeadm/issues/129

**Special notes for your reviewer**:

This PR is should make it possible to secure the apiserver -> kubelet communication.

**Release note**:

```release-note
NONE
```

@pires @mikedanese @andrewrynhard @liggitt @deads2k

### CODE REVIEW

File: cmd/kubeadm/app/cmd/init.go [hunk 1]
```diff
@@ -168,12 +168,6 @@ func NewInit(cfgPath string, cfg *kubeadmapi.MasterConfiguration, skipPreFlight
 	// Try to start the kubelet service in case it's inactive
 	preflight.TryStartKubelet()
 
-	// Warn about the limitations with the current cloudprovider solution.
-	if cfg.CloudProvider != "" {
-		fmt.Println("WARNING: For cloudprovider integrations to work --cloud-provider must be set for all kubelets in the cluster.")
-		fmt.Println("\t(/etc/systemd/system/kubelet.service.d/10-kubeadm.conf should be edited for this purpose)")
-	}
-
 	return &Init{cfg: cfg, selfHosted: selfHosted}, nil
```
Comment (pires) [inline cmd/kubeadm/app/cmd/init.go]: Why is this removed?

File: cmd/kubeadm/app/cmd/join.go [hunk 2]
```diff
@@ -136,10 +138,20 @@ func (j *Join) Run(out io.Writer) error {
...
+
+	kubeconfigFile := filepath.Join(kubeadmapi.GlobalEnvParams.KubernetesDir, kubeconfigphase.KubeletKubeConfigFileName)
+	if err := kubeconfigphase.WriteKubeconfigToDisk(kubeconfigFile, cfg); err != nil {
 		return err
 	}
 
+	// Write the ca certificate to disk so kubelet can use it for authentication
+	cluster := cfg.Contexts[cfg.CurrentContext].Cluster
+	caCertFile := filepath.Join(kubeadmapi.GlobalEnvParams.KubernetesDir, kubeadmconstants.CACertName)
+	err = certutil.WriteCert(caCertFile, cfg.Clusters[cluster].CertificateAuthorityData)
```
Comment (pires) [inline cmd/kubeadm/app/cmd/join.go]: Where will we be using the CA file written here?

**Review by pires — CHANGES_REQUESTED** [review]
Couple questions but the code itself looks good.

File: cmd/kubeadm/app/cmd/init.go (hunk 1 above)
Comment (luxas) [inline cmd/kubeadm/app/cmd/init.go]: Cleanup. It's covered in defaults.go

File: cmd/kubeadm/app/cmd/join.go (hunk 2 above)
Comment (luxas) [inline cmd/kubeadm/app/cmd/join.go]: In the debs/rpms with the --client-ca-cert file, see the PR description above

**Review by pires — APPROVED** [review]

**Comment by luxas** [comment]
@k8s-bot kops aws e2e test this

File: cmd/kubeadm/app/phases/certs/certs.go [hunk 3]
```diff
@@ -123,7 +123,9 @@ func CreatePKIAssets(cfg *kubeadmapi.MasterConfiguration, pkiDir string) error {
 		config := certutil.Config{
 			CommonName: "kube-apiserver",
 			AltNames:   altNames,
-			Usages:     []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+			// This makes the apiserver allowed to talk to the kubelets in the cluster
+			Organization: []string{"system:masters"},
```
Comment (liggitt) [inline cmd/kubeadm/app/phases/certs/certs.go]: this doesn't belong in the serving cert, it's meaningless there. you want this in the client cert you need to make for the master

Comment (luxas) [inline cmd/kubeadm/app/phases/certs/certs.go]: Yeah, I'm gonna fix that. Misunderstood which way you meant it should be created, but it's clearer now, thank you.

Comment (luxas) [inline cmd/kubeadm/app/phases/certs/certs.go]: @liggitt Do you think something like this will work then?
```
--- a/cmd/kubeadm/app/constants/constants.go
+++ b/cmd/kubeadm/app/constants/constants.go
@@ -24,4 +24,8 @@ const (
        APIServerCertAndKeyBaseName = "apiserver"
        APIServerCertName           = "apiserver.crt"
        APIServerKeyName            = "apiserver.key"
+
+       APIServerKubeletClientCertAndKeyBaseName = "apiserver-kubelet-client"
+       APIServerKubeletClientCertName           = "apiserver-kubelet-client.crt"
+       APIServerKubeletClientKeyName            = "apiserver-kubelet-client.key"
 )
diff --git a/cmd/kubeadm/app/master/manifests.go b/cmd/kubeadm/app/master/manifests.go
index bf44fad..b234b70 100644
--- a/cmd/kubeadm/app/master/manifests.go
+++ b/cmd/kubeadm/app/master/manifests.go
@@ -322,8 +322,8 @@ func getAPIServerCommand(cfg *kubeadmapi.MasterConfiguration, selfHosted bool) [
                "--client-ca-file="+getCertFilePath(kubeadmconstants.CACertName),
                "--tls-cert-file="+getCertFilePath(kubeadmconstants.APIServerCertName),
                "--tls-private-key-file="+getCertFilePath(kubeadmconstants.APIServerKeyName),
-               "--kubelet-client-certificate="+getCertFilePath(kubeadmconstants.APIServerCertName),
-               "--kubelet-client-key="+getCertFilePath(kubeadmconstants.APIServerKeyName),
+               "--kubelet-client-certificate="+getCertFilePath(kubeadmconstants.APIServerKubeletClientCertName),
+               "--kubelet-client-key="+getCertFilePath(kubeadmconstants.APIServerKubeletClientKeyName),
                "--token-auth-file="+kubeadmapi.GlobalEnvParams.HostPKIPath+"/tokens.csv",
                fmt.Sprintf("--secure-port=%d", cfg.API.Port),
                "--allow-privileged",
diff --git a/cmd/kubeadm/app/phases/certs/certs.go b/cmd/kubeadm/app/phases/certs/certs.go
index 78d5533..4ac2a13 100644
--- a/cmd/kubeadm/app/phases/certs/certs.go
+++ b/cmd/kubeadm/app/phases/certs/certs.go
@@ -123,8 +123,6 @@ func CreatePKIAssets(cfg *kubeadmapi.MasterConfiguration, pkiDir string) error {
                config := certutil.Config{
                        CommonName: "kube-apiserver",
                        AltNames:   altNames,
-                       // This makes the apiserver allowed to talk to the kubelets in the cluster
-                       Organization: []string{"system:masters"},
                        Usages:       []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
                }
                apiCert, apiKey, err := pkiutil.NewCertAndKey(caCert, caKey, config)
@@ -138,6 +136,35 @@ func CreatePKIAssets(cfg *kubeadmapi.MasterConfiguration, pkiDir string) error {
                fmt.Println("[certificates] Generated API server certificate and key.")
        }
 
+       // If at least one of them exists, we should try to load them
+       // In the case that only one exists, there will most likely be an error anyway
+       if pkiutil.CertOrKeyExist(pkiDir, kubeadmconstants.APIServerKubeletClientCertAndKeyBaseName) {
+               // Try to load ca.crt and ca.key from the PKI directory
+               apiCert, apiKey, err := pkiutil.TryLoadCertAndKeyFromDisk(pkiDir, kubeadmconstants.APIServerKubeletClientCertAndKeyBaseName)
+               if err != nil || apiCert == nil || apiKey == nil {
+                       return fmt.Errorf("certificate and/or key existed but they could not be loaded properly")
+               }
+
+               fmt.Println("[certificates] Using the existing API Server certificate and key.")
+       } else {
+               // The certificate and the key did NOT exist, let's generate them now
+               // TODO: Add a test case to verify that this cert has the x509.ExtKeyUsageClientAuth flag
+               config := certutil.Config{
+                       CommonName: "kube-apiserver-kubelet-client",
+                       Organization: []string{"system:masters"},
+                       Usages:       []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
+               }
+               apiClientCert, apiClientKey, err := pkiutil.NewCertAndKey(caCert, caKey, config)
+               if err != nil {
+                       return fmt.Errorf("failure while creating API server client key and certificate [%v]", err)
+               }
+
+               if err = pkiutil.WriteCertAndKey(pkiDir, kubeadmconstants.APIServerKubeletClientCertAndKeyBaseName, apiClientCert, apiClientKey); err != nil {
+                       return fmt.Errorf("failure while saving API server client certificate and key [%v]", err)
+               }
+               fmt.Println("[certificates] Generated API server certificate and key.")
+       }
+
        fmt.Printf("[certificates] Valid certificates and keys now exist in %q\n", pkiDir)
 
        return nil
```

Comment (liggitt) [inline cmd/kubeadm/app/phases/certs/certs.go]: at first glance, that seems more correct
